src/Contract/StartBidding.purs

```diff
@@ -0,0 +1,302 @@
+  )
+
+newtype StartBiddingContractParams = StartBiddingContractParams
+  { auctionInfo :: AuctionInfo
+  }
+
+derive instance Generic StartBiddingContractParams _
+derive instance Newtype StartBiddingContractParams _
+derive instance Eq StartBiddingContractParams
+
+instance Show StartBiddingContractParams where
+  show = genericShow
+
+instance HasJson StartBiddingContractParams where
+  jsonCodec = const startBiddingContractParamsCodec
+
+startBiddingContractParamsCodec :: CA.JsonCodec StartBiddingContractParams
+startBiddingContractParamsCodec =
+  wrapIso StartBiddingContractParams $ CA.object "StartBiddingContractParams" $
+    CAR.record
+      { auctionInfo: auctionInfoCodec
+      }
+
+startBiddingContract
+  :: StartBiddingContractParams
+  -> Contract (ContractOutput TransactionHash)
+startBiddingContract =
+  mkContractOutput _.txHash <<< mkStartBiddingContractWithErrors
+
+mkStartBiddingContractWithErrors
+  :: StartBiddingContractParams
+  -> ExceptT StartBiddingContractError Contract ContractResult
+mkStartBiddingContractWithErrors (StartBiddingContractParams params) = do
+  let
+    auctionInfo@(AuctionInfo auctionInfoRec) = params.auctionInfo
+    auctionTerms@(AuctionTerms auctionTermsRec) = auctionInfoRec.auctionTerms
+
+  -- Check auction terms:
+  validateAuctionTerms auctionTerms #
+    validation (throwError <<< StartBidding_Error_InvalidAuctionTerms) pure
+
+  -- Check that the contract is initiated by the seller:
+  ownPkh <- ownPaymentPubKeyHash !? StartBidding_Error_CouldNotGetOwnPubKeyHash
+  when (unwrap ownPkh /= auctionTermsRec.sellerPkh) $
+    throwError StartBidding_Error_ContractNotInitiatedBySeller
+
+  -- Check that the current time is within the bidding period:
+  nowTime <- lift currentTime
+  when (nowTime < auctionTermsRec.biddingStart) $
+    throwError StartBidding_Error_CurrentTimeBeforeBiddingStart
+  when (nowTime >= auctionTermsRec.biddingEnd) $
+    throwError StartBidding_Error_CurrentTimeAfterBiddingEnd
+
+  -- Build auction escrow validator:
+  auctionEscrowValidator <-
+    withExceptT StartBidding_Error_CouldNotBuildAuctionEscrowValidator $
+      mkAuctionEscrowValidatorFromAuctionInfo auctionInfo
+  let
+    auctionEscrowValidatorHash = validatorHash auctionEscrowValidator
+    auctionEscrowValidatorAddress = scriptHashAddress auctionEscrowValidatorHash Nothing
+
+  -- Check that the computed auction escrow validator address matches
+  -- the address specified in auction info:
+  when (auctionEscrowValidatorAddress /= auctionInfoRec.auctionEscrowAddr) $
+    throwError StartBidding_Error_AuctionEscrowValidatorAddressMismatch
+
+  -- Get standing bid validator hash:
+  standingBidValidatorHash <- addressPaymentValidatorHash auctionInfoRec.standingBidAddr
+    ?? StartBidding_Error_CouldNotGetStandingBidValidatorHash
+
+  -- Query current auction escrow utxo:
+  auctionEscrowUtxo <- queryAuctionEscrowUtxo auctionInfo
+    !? StartBidding_Error_CouldNotFindCurrentAuctionEscrowUtxo
```

Comment:
It's good that you build the auctionEscrowValidatorAddress yourself and compare it against the one declared in the auctionInfo.

I think you should also build the standingBidValidatorAddress yourself and compare it against the one in auctionInfo.

In general, our auction minting policy doesn't prevent invalid addresses from being set in auctionInfo, so we must re-verify those addresses ourselves in every off-chain endpoint. Otherwise, a malicious auctionInfo record could cause funds to be sent to the wrong places or misleading information to be displayed as auction state.

Practically, this may not be needed in simple queries like "get all auctions" as it would probably slow down the query too much. However, I think it's necessary for any endpoints that interact with auction state on the user's behalf.

Later on in the project, if we decide to trust auction info for simple queries, we may want to implement some more efficient way to batch-process auction info and generate a blocklist to hide auctions with invalid auctionInfo from the UI.

Anyway, for now, please build and verify all auction script addresses when building transactions.

Reply:
Btw, how long does it take to build the validator addresses? Is it something that can be used in queries without slowing them down too much -- perhaps just for single-auction state but not multi-auction summary?

Reply:
Partially addressed in https://github.com/mlabs-haskell/hydra-auction-offchain/pull/15 (caching will be introduced in a separate PR)